File: packages/urdu-dotnet/UrduTools.Core/Normalization/Fingerprint.cs
namespace UrduTools.Core.Normalization;

public static class Fingerprint
{
    private static readonly NormalizeOptions FingerprintOptions = new()
    {
        Nfc = true, Nbsp = true, AlifMadda = true, Numerals = true,
        ZeroWidth = true, Diacritics = true, Honorifics = true, Hamza = true,
        Kashida = false, PresentationForms = false, PunctuationTrim = true,
        NormalizeCharacters = false,
    };

    public static string Compute(string text)
    {
        if (string.IsNullOrEmpty(text)) return string.Empty;
        return UrduNormalizer.Normalize(text, FingerprintOptions).Trim();
    }
}

File: packages/urdu-dart/lib/src/normalization/fingerprint.dart
import '../types.dart';
import 'normalize.dart';

/// Produces a canonical comparison fingerprint for an Urdu word.
///
/// Combines normalization layers 1-8 plus punctuation trimming.
/// Designed for client-side word-status matching without a DB round-trip.
String fingerprint(String text) {
  if (text.isEmpty) return '';
  return normalize(
      text,
      const NormalizeOptions(
        nfc: true,
        nbsp: true,
        alifMadda: true,
        numerals: true,
        zeroWidth: true,
        diacritics: true,
        honorifics: true,
        hamza: true,
        kashida: false,
        presentationForms: false,
        punctuationTrim: true,
      )).trim();
}
